### PR TITLE
[C] C generator refactored

### PR DESCRIPTION
**Pull request by zhemant**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Updated mustache for model and api to cover as much possible data types
Updated samples according to new mustaches
change data type from file to binary_t to handle binary data (support for OAS3)
Updated cmake mustache

@wing328

### CODE REVIEW

File: samples/client/petstore/c/unit-tests/manual-PetAPI.c [hunk 1]
```diff
@@ -115,15 +112,26 @@ int main() {
...
 		                                          EXAMPLE_PET_ID,
 		                                          "dec",
-		                                          file);
+		                                          data);
 
 		api_response_free(respo);
 		fclose(file);
 	} else {
-    apiClient_free(apiClient3);
+    //apiClient_free(apiClient3);
```
Comment (wing328) [inline samples/client/petstore/c/unit-tests/manual-PetAPI.c]: I would suggest removing this line as it's commented out and people have no idea why it's left there.

Comment (zhemant) [reply samples/client/petstore/c/unit-tests/manual-PetAPI.c]: I updated it, as apiClient object has to be freed once the requirement is over. Modified all test samples accordingly.

File: samples/client/petstore/c/unit-tests/manual-UserAPI.c [hunk 2]
```diff
@@ -102,16 +105,19 @@ int main() {
...
 
 	UserAPI_logoutUser(apiClient4);
+    apiClient_free(apiClient4);
 
 
 // delete user test
 	apiClient_t *apiClient5 = apiClient_create();
 
 	UserAPI_deleteUser(apiClient5, "example123");
+    apiClient_free(apiClient5);
```
Comment (wing328) [inline samples/client/petstore/c/unit-tests/manual-UserAPI.c]: Minor suggestion: indent ` apiClient_free` properly with reference to the previous line.

Comment (wing328) [reply samples/client/petstore/c/unit-tests/manual-UserAPI.c]: You may simply want to use the code formatter to format the manually written test files.

Comment (zhemant) [reply samples/client/petstore/c/unit-tests/manual-UserAPI.c]: I have formatted manually written test files with uncrustify with the configuration we have in auto-generated code.

**Comment by wing328** [comment]
@zhemant what about using FILE* instead? 

It seems to me FILE* is pretty common for handling files in C.

**Comment by zhemant** [comment]
Yes, but this new structure is keeping in mind we also need to handle binary data which may not be a file pointer. In OAS3 binary and file are the same. And in curl, we need to pass the binary/file as data(char*/uint8_t) and its size. Hence to make it common for both data type user has to pass a struct of binary_t with data and length of data so it is easy to handle both files and binary while sending.

In current codes for handling FILE* we are doing same, reading data into a struct with char* and length.

**Comment by wing328** [comment]
@zhemant thanks for the explanation, which totally makes sense.

**Comment by wing328** [comment]
## Upgrade Note

Binary and file are now handled with the binary_t pointer

**Pull request merged**